File: Day01/C++/Quentin/main.cpp

#ifndef QUENTIN
    #define QUENTIN
    #define NAME "Quentin"
    #define COOKIE "53616c7465645f5fb0cb7b5bb9806356f97ca289cd994f856eeab93000a4133b3bf39c7e1d7988e87373986e95dca6170f978d0e297de240b4709295aee49d0a"
#endif

#include "limits.h"
#include "my_lib.hpp"

int main() {
    std::vector<std::string> input = get_input();

    std::vector<int> left;
    std::vector<int> right;
    std::unordered_map<int, int> counts;

    for(auto line : input) {
        int firstNbr = atoi(line.c_str());
        int nextIndex = 2;
        if (firstNbr != 1) {
            nextIndex += (int)std::ceil(std::log10(firstNbr));
        } else {
            nextIndex++;
        }
        int sndNbr = atoi(line.c_str() + nextIndex);
        left.push_back(firstNbr);
        right.push_back(sndNbr);

        counts[sndNbr]++;
    }

    left = sort(left);
    right = sort(right);

    auto len = left.size();
    size_t sum = 0;
    for (int i = 0; i < len; i++) {
        sum += std::abs(left[i] - right[i]);
    }
    std::cout << "Sum: " << sum << std::endl;

    size_t similarity = 0;
    for (auto item : left) {
        similarity += item * counts[item];
    }

    std::cout << "Similarity: " << similarity << std::endl;
}

File: Day01/Rust/Quentin/main.rs
use std::fs::File;
use std::io::BufReader;
use std::io::prelude::*;

fn my_atoi(s: &str, start: usize) -> i32 {
    let mut i = start;
    let mut result = 0;
    let mut sign = 1;
    if s.as_bytes()[i] == 45 {
        sign = -1;
        i += 1;
    }
    while i < s.len() && s.as_bytes()[i] >= 48 && s.as_bytes()[i] <= 57 {
        result = result * 10 + (s.as_bytes()[i] - 48) as i32;
        i += 1;
    }
    return result * sign;
}

fn main() -> std::io::Result<()> {
    let file = File::open("./resources/Day01/Quentin.txt")?;
    let mut buf_reader = BufReader::new(file);
    let mut contents = String::new();
    buf_reader.read_to_string(&mut contents)?;

    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    let mut counts: std::collections::HashMap<i32, i32> = std::collections::HashMap::new();

    for line in contents.lines() {
        let first_number = my_atoi(line, 0);
        let mut next_index = 3;
        if first_number != 1 {
            next_index += (first_number as f64).log10().ceil() as usize;
        } else {
            next_index += 1;
        }
        let second_number = my_atoi(line, next_index);
        
        left.push(first_number);
        right.push(second_number);

        let count = counts.entry(second_number).or_insert(0);
        *count += 1;
    }

    left.sort();
    right.sort();

    let mut sum = 0;
    for i in 0..left.len() {
        sum += (left[i] - right[i]).abs();
    }
    println!("Sum: {}", sum);

    let mut similarity = 0;
    for i in 0..left.len() {
        if let Some(count) = counts.get(&left[i]) {
            similarity += left[i] * count;
        }
    }
    println!("Similarity: {}", similarity);

    Ok(())
}